src/hotspot/share/memory/memoryReserver.cpp

```diff
@@ -549,8 +549,17 @@ ReservedHeapSpace HeapReserver::Instance::reserve_compressed_oops_heap(const siz
   const size_t attach_point_alignment = lcm(alignment, os_attach_point_alignment);
 
   char* aligned_heap_base_min_address = align_up((char*)HeapBaseMinAddress, alignment);
-  size_t noaccess_prefix = ((aligned_heap_base_min_address + size) > (char*)OopEncodingHeapMax) ?
-    noaccess_prefix_size : 0;
+  char* heap_end_address = aligned_heap_base_min_address + size;
+
+  bool unscaled  = false;
+  bool zerobased = false;
+  bool heapbased = UseCompatibleCompressedOops;
```

Comment:
Use of `heapbased` is a bit confusing to me. Can we restructure it like this:
```
bool unscaled = false;
bool zerobased = false;
if (UseCompatibleCompressedOops) {
  unscaled = (heap_end_address <= (char*)UnscaledOopHeapMax);
  zerobased = (heap_end_address <= (char*)OopEncodingHeapMax);
}
size_t noaccess_prefix = (UseCompatibleCompressedOops || !zerobased) ? noaccess_prefix_size : 0;
```

Reply:
I can do that. I got names from [CompressedOops::Mode](https://github.com/openjdk/leyden/blob/premain/src/hotspot/share/oops/compressedOops.hpp#L67) values